lib/Sema/TypeCheckType.cpp

```diff
@@ -1996,7 +1996,27 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
       hasFunctionAttr = true;
       break;
     }
-
+  
+  // If we have an @autoclosure then try resolving the top level type repr
+  // first as it may be pointing to a typealias
+  if (attrs.has(TAK_autoclosure)) {
+    if (auto CITR = dyn_cast<ComponentIdentTypeRepr>(repr)) {
+      auto typeAliasResolver = TypeResolverContext::TypeAliasDecl;
+      if (auto type = resolveTopLevelIdentTypeComponent(resolution, CITR,
+                                                        typeAliasResolver)) {
+        if (auto TAT = dyn_cast<TypeAliasType>(type.getPointer())) {
+          if (auto underlyingRepr = TAT->getDecl()
+                                       ->getUnderlyingTypeLoc()
+                                        .getTypeRepr()) {
+            if (!underlyingRepr->isInvalid()) {
```

Comment:
One question - since `resolveTopLevelIdentTypeComponent` supposed to return a valid type or `ErrorType` if there is a problem - could `isInvalid()` ever be `true` here?

Reply:
It seems like two last `if`s could be simplified into `repr = TAT->getDecl()->getUnderlyingTypeLoc().getTypeRepr();` since the type originated from repr in this case.

Reply:
Yeah, I was just writing the code a bit defensively! I have ran another validation test and there wasn't any issue so I have now gotten rid of the extra `if` checks. 

Could you run the CI for me? :)

Reply:
@xedin Yeah, I was just writing the code a bit defensively! I have ran another validation test and there wasn't any issue so I have now gotten rid of the extra `if` checks. 

Could you run the CI for me? :)